lib/bolt/plugin/azure.rb

```diff
@@ -0,0 +1,200 @@
+# frozen_string_literal: true
+
+require 'json'
+
+module Bolt
+  class Plugin
+    class Azure
+      class AzureHTTPError < Bolt::Error
+        def initialize(response)
+          err = JSON.parse(response.body).dig('error', 'message')
+          m = String.new("#{response.code} \"#{response.msg}\"")
+          m += ": #{err}" if err
+          super(m, 'bolt.plugin/azure-http-error')
+        end
+      end
+
+      attr_reader :config
+
+      CONFIG_KEYS = Set['tenant_id', 'client_id', 'client_secret', 'subscription_id', 'profile']
+      OPTS_KEYS = Set['_plugin', 'location', 'tags', 'resource_group', 'scale_set'] + CONFIG_KEYS
+
+      def initialize(config)
+        validate_config(config)
+        @config = config
+        @logger = Logging.logger[self]
+      end
+
+      def name
+        'azure'
+      end
+
+      def hooks
+        %w[inventory_targets]
+      end
+
+      def validate_config(config)
+        keys = config.keys.to_set
+
+        unless CONFIG_KEYS.superset?(keys)
+          keys -= CONFIG_KEYS
+          raise Bolt::ValidationError, "Unexpected key(s) in plugin config: #{keys.to_a.inspect}"
+        end
+      end
+
+      def validate_options(opts)
+        keys = opts.keys.to_set
+
+        unless OPTS_KEYS.superset?(keys)
+          keys -= OPTS_KEYS
+          raise Bolt::ValidationError, "Unexpected key(s) in inventory config: #{keys.to_a.inspect}"
+        end
+      end
+
+      def inventory_targets(opts)
+        validate_options(opts)
+        creds = credentials(opts)
+        token = token(creds)
+        instances = instances(token, creds, opts)
+
+        # Filter by location
+        if opts['location']
+          instances.select! do |instance|
+            instance['location'].eql? opts['location']
```

Comment:
Is there a reason this is `eql?` instead of `==`?

Reply:
Oops. Just a force of habit from grading/writing projects in C... Updated. :)